### ISSUE TITLE
FuelRemaining incorrect on MX-30

### ISSUE THREAD

**Issue by Frankenslag** [bdr99/pymazda#14]
I have noticed that the **FuelDistanceRemainingKm** is being reported incorrectly for my MX-30. (120km instead of 218km). After experimentation seems that there is an additional API, **getEVVehicleStatus**, that can return the correct value along with the charge percentage etc. I believe you can if you have an electric car by examining if **eConnectType** in the vehicle base info is set to 1.

**Comment by bdr99** [comment]
Thanks for this info! I don't have an MX-30 so this is very helpful. Can you provide a sample of the response for getEVVehicleStatus so I can see the structure of the data?

**Comment by Frankenslag** [comment]
See below for an example JSON response 

I looked for strings in the apk similar to the ones used in your library and found the **getEVVehicleStatus** close by. I didn't really do much analysis in the traditional sense. The interesting fields for me are **SmaphRemDrvDistKm**, **SmaphRemDrvDistMile** and **SmaphSOC**. I would not use the position figures returned by this as they don't have the sign correct and I don't see anything to indicate the correct sign.

```
{
   "resultData":[
      {
         "NId":"FFFF",
         "InformationDatetime":"2021-08-07T08:40:26Z",
         "PlusBInformation":{
            "VehicleInfo":{
               "ChargeInfo":{
                  "SmaphRemDrvDistKm":218,
                  "SmaphSOC":100,
                  "CstmzStatBatHeatAutoSW":1,
                  "SmaphRemDrvDistMile":136,
                  "ACChargeStatus":0,
                  "ChargeStatusSub":0,
                  "ChargerConnectorFitting":0,
                  "MaxChargeMinuteAC":0,
                  "BatteryHeaterON":0,
                  "DCChargeStatus":0,
                  "MaxChargeMinuteQBC":0
               },
               "RemoteHvacInfo":{
                  "HVAC":0,
                  "RearDefogger":0,
                  "InCarTeDC":15.1,
                  "InCarTeDF":59.18,
                  "FrontDefroster":0
               }
            }
         },
         "NotificationCategory":36,
         "SId":"FFFF",
         "OccurrenceTime":"2021-08-07T08:40:26Z",
         "DcmNumber":"9999999999900000",
         "BsId":"FFFF",
         "DcmDormantDatetime":"2021-08-07T15:10:43Z",
         "IgInformation":{
            "VehicleInfo":{
               "ChargeInfo":{
                  "ChargeScheduleStatus":1048575,
                  "LastUpdatedTimeForScheduledChargeTime":"2021-08-11T08:31:42Z"
               }
            }
         },
         "OccurrenceDate":"20210807083956",
         "PositionInfo":{
            "AcquisitionDatetime":"2021-07-29T10:58:22Z",
            "Latitude":xx.xxxxxxx,
            "Longitude":x.xxxxxxx,
            "DcmPositionAccuracy":{
               "Gradient":45,
               "MinorAxisError":2,
               "AcquisitionState":1,
               "MajorAxisError":2
            }
         },
         "PositionInfoCategory":5,
         "TransmissionFactor":"32"
      }
   ],
   "resultCode":"200S00",
   "visitNo":"xxxxxxxxxxxxxxxxxxxx"
}
```

**Comment by bdr99** [comment]
Thanks. What is `SmaphSOC`? Is that the battery level as a percentage?

**Comment by Frankenslag** [comment]
Another end point that may be useful is **getRemoteControlPermission**. I believe that you can restrict the API to certain functions and this will tell you what is available. It is called with the vin rather than the internal vin as below

```
 {
   "internaluserid":"__INTERNAL_ID__",
   "vin":"xxxxxxxxxxxxxxxxx"
}

```

The response is :

```
{
   "remoteControl":{
      "eConnectRemoteHvac":{
         "airConditioningOn":1,
         "airConditioningOff":1
      },
      "charge":{
         "immediatelyCharge":1,
         "stopPowerSupply":1
      },
      "doorLock":{
         "lockBaguraSetOff":1,
         "lockBaguraSetOn":1,
         "unLock":1,
         "remoteLockCancel":1
      },
      "hazard":{
         "hazardOffCancel":0,
         "carFinder":0,
         "hazardOff":0
      },
      "remoteHvac":{
         "engineOff":1,
         "engineOn":1
      }
   },
   "resultCode":"200S00",
   "visitNo":"xxxxxxxxxxxxxxxxxxxxxx"
}
```

**Comment by Frankenslag** [comment]
> Thanks. What is `SmaphSOC`? Is that the battery level as a percentage?

I believe so - will let you know the next time someone actually drives the car and I can compare readings.

**Comment by bdr99** [comment]
Thanks, this is some great info. There are so many endpoints that I haven't figured out yet. I'll start by adding the EV status attributes to the return value of `get_vehicle_status()`. Would you be willing to help me research other EV-specific methods as well, such as the HVAC methods?

**Comment by Frankenslag** [comment]
> Thanks, this is some great info. There are so many endpoints that I haven't figured out yet. I'll start by adding the EV status attributes to the return value of `get_vehicle_status()`. Would you be willing to help me research other EV-specific methods as well, such as the HVAC methods?

Of course - Some background.. I have an MX5 (Miata) RF and an MX-30.

I have been working on a port of your library to C# which is almost finished. Happy to share this library with you if you want to keep it alongside your python version..

**Comment by bdr99** [comment]
The C# port sounds interesting. If you have any questions about my Python or TypeScript versions feel free to ask.

Are you planning to open source it on GitHub? I'd be interested to see how it looks.

**Comment by Frankenslag** [comment]
> The C# port sounds interesting. If you have any questions about my Python or TypeScript versions feel free to ask.
> 
> Are you planning to open source it on GitHub? I'd be interested to see how it looks.

I will Open Source it once I write a readme. I have added you as a collaborator so you can see it now at https://github.com/Frankenslag/MazdaApi

**Comment by bdr99** [comment]
As an update, I'm still working on adding the `getEVVehicleStatus` endpoint, but I want to figure out what more of the values mean first. Working on debugging the apk to figure them out.

**Comment by Frankenslag** [comment]
I have a further api that I have looked at on my MX-30. This is Get & Set HVAC

TemperatureType = 1 for Celsius and 2 For Farenheit 

**End Point**

**/remoteServices/updateHVACSetting/v4**

**Body**

```
{
   "internaluserid":"__INTERNAL_ID__",
   "internalvin":xxxxxxxxx,
   "hvacsettings":{
      "FrontDefroster":0,
      "RearDefogger":0,
      "Temperature":20.0,
      "TemperatureType":1
   }
}
```

**Response**

```
{
   "resultCode":"200S00",
   "message":"[English (UK)]The request was completed successfully. (200S00)",
   "visitNo":"GWxxxxxxxxxxxxxxxxxxxx"
}
```

**End Point**

/remoteServices/getHVACSetting/v4

**Body**

```
{
   "internaluserid":"__INTERNAL_ID__",
   "vin":"xxxxxxxxxx"
}
```

**Response**

```
{
   "resultCode":"200S00",
   "hvacSettings":{
      "TemperatureType":1,
      "Temperature":20.0,
      "RearDefogger":0,
      "FrontDefroster":0
   },
   "visitNo":"GWxxxxxxxxxxxxxxxxxxxx"
}
```

**Comment by Frankenslag** [comment]
I have made the C# version publicly available [Here](https://github.com/Frankenslag/MazdaApi)

**Comment by bdr99** [comment]
@Frankenslag Thanks for the info about `updateHVACSetting` and `getHVACSetting`. Would you mind sharing sample responses for the following endpoints as well?

 - `remoteServices/activeRealTimeVehicleStatus/v4`
 - `/remoteServices/findVehicleLocation/v4`
 - `/remoteServices/hvacOn/v4`
 - `/remoteServices/hvacOff/v4`

They all take a request body that looks like this:
```json
{
   "internaluserid":"__INTERNAL_ID__",
   "internalvin": 12345
}
```

These are all specific to the MX-30 so I'm unable to test them myself. Of course make sure to censor any personal information in the responses.

**Comment by Frankenslag** [comment]
For both **remoteServices/hvacOn/v4** and **remoteServices/hvacOff/v4** the result was in the following format. 

```
{
   "resultCode":"200S00",
   "message":"[English (UK)]The request was completed successfully. (200S00)",
   "visitNo":"GWxxxxxxxxxxxxxxxxxxxx"
}
```

**Comment by Frankenslag** [comment]
Here is a response to **remoteServices/findVehicleLocation/v4**. Note that the flags need to be used to determine the sign of the Lattitude and Longitude data.

```
{
   "resultCode":"200S00",
   "AcquisitionDatetime":"20210101100000",
   "Latitude":12.345678,
   "LatitudeFlag":0,
   "LongitudeFlag":0,
   "Longitude":1.345678,
   "visitNo":"GWxxxxxxxxxxxxxxxxxxxx"
}
```

**Comment by bdr99** [comment]
Hi @Frankenslag I added some new EV-specific features to this library:
 - `isElectric` attribute in the response of `get_vehicles()`
 - `get_ev_vehicle_status()` - this has the range info, battery status, etc
 - `get_hvac_setting()`
 - `set_hvac_setting()`
 - `turn_on_hvac()`
 - `turn_off_hvac()`
 - `refresh_vehicle_status()`

I pushed the changes to a branch named [`ev_hvac`](https://github.com/bdr99/pymazda/tree/ev_hvac). Whenever you get a chance, can you please test these features on your car? They are all documented in the readme on that branch.

I have a couple of questions in particular:
 - If you call `set_hvac_setting()` when the HVAC is off, are the new HVAC settings updated on the car? Or do you need to turn the HVAC on in order for the new settings to appear?
 - `get_ev_vehicle_status()` as well as `get_hvac_setting()` both include info about whether the front and rear defrosters are enabled. I'm not sure why this information is repeated in two different endpoints. Can you check if both of them are accurate?
 - I think that `remoteServices/activeRealTimeVehicleStatus/v4` might be an endpoint to request a new status update from the car. If you call this method, then the `lastUpdatedTimestamp` property from `get_vehicle_status()` and `get_ev_vehicle_status()` should update, and the response should reflect the latest status info. Can you confirm this? This is the `refresh_vehicle_status()` method in the library.

Thanks in advance!

**Comment by CacioGit** [comment]
I just started to drive an MX-30, and noticed that the main branch doesn't seem to include the EV-specific methods.
On the other side, the ev_hvac branch is now a couple of releases outdated.

Do you think you could merge at least the get_ev_vehicle_status() method into the main branch?

Thank you so much for all your work on this BTW

**Comment by bdr99** [comment]
Thanks @CacioGit for pointing that out. You're right, the ev_hvac branch is pretty outdated at this point. I went ahead and merged it into the main branch and released it as 0.3.0. Hope this helps!

If you wouldn't mind, could you please test the EV-specific features I mentioned above and let me know if they are working correctly with your car? I don't have an MX-30 so I can't test them myself. Thanks!

**Comment by CacioGit** [comment]
Sorry for the delayed response..
I immediately downloaded your 0.3.0 release, and indeed the new get_ev_vehicle_status() and refresh_vehicle_status() work perfectly.

I did some experimenting with retrieving the ev_vehicle_status() with and without a call to refresh(), and I think I started to understand the logic behind what the MX will actually update, and what not..

Both from the actual values themselves, and also from the various "timestamp" entries, it is really the case that a refresh() will ONLY refresh the ev_vehicle values, and NOT the more generic get_vehicle_status() ones.

This has the side effect of seeing brand new values for the battery details, but possibly quite old ones on the odometerKm or GPS position.

I could confirm that even the official Mazda app (iOS), when you ask for a refresh, you will get new data (with a current timestamp) for the battery charge status, or the car heating.
On the other hand, the odometer, and the tire pressure, isn't refreshed:
the app actually shows the correct, older, timestamp associated with those values.

I think this is also an answer to the other open issue "Call for Update "lastUpdatedTimestamp":
your API/library, is already doing everything what the Mazda app also does in respect to what is "real time", and what is not.

The generic get_vehicle_status() values seem to be updated at the end of each trip (so when the car is switched off), but also when a battery charge session is complete, probably because at that point in time the car also goes "to sleep", and sends an update to the central Mazda servers before powering itself down.

I will try to go over the questions in the questions on the *_hvac_setting() over the next few days and will let you know what I find out.
